aesm-client/src/imp/aesm_protobuf/mod.rs

```diff
@@ -107,4 +109,150 @@ impl AesmClient {
+    ///
+    /// @returns structure containing set of keys that can serve as input for init_quote_ex/get_quote_size_ex operations. (see AesmKeyIds::select_algorithm_id to extract a specific key)
+    ///
+    pub fn get_supported_att_key_ids(&self) -> Result<AesmKeyIds> {
+        let mut req = Request_GetSupportedAttKeyIDsRequest::new();
+
+        // FIXME: remove conditional compilation after resolving https://github.com/fortanix/rust-sgx/issues/31
+        #[cfg(not(target_env = "sgx"))]
+        req.set_timeout(REMOTE_AESM_TIMEOUT_US);
+
+        // sgx defined protobuf returns an opaque byte array with size we give it. Byte array MUST have expected size of structure size*num.
+        let num_keys : u32 = self.get_supported_att_key_id_num()?;
+        let expected_buffer_size : u32 = num_keys * (::std::mem::size_of::<sgx_att_key_id_ext_t>() as u32);
+        req.set_buf_size(expected_buffer_size);
+
+        // fetch requested data
+        let mut res = self.transact(req)?;
+
+        // Interpret the byte array as defined in C header at linux-sgx/common/inc/sgx_quote.h, this is converted to sgx_quote.rs via bindgen.
+        let keys = res.take_att_key_ids();
+
+        if keys.len() as u32 != expected_buffer_size {
+            return Err(Error::aesm_bad_response("wrong buffer size returned by aesm service"));
+        }
+
+        Ok(AesmKeyIds { keys } )
+    }
+    
+    ///
+    /// Returns information needed by an intel SGX application to get a quote of one of its enclaves.
+    /// Similar functionality to sgx_init_quote_ex in page 165 at https://download.01.org/intel-sgx/sgx-linux/2.9.1/docs/Intel_SGX_Developer_Reference_Linux_2.9.1_Open_Source.pdf
+    ///
+    /// @param[in] att_key_id   - a byte array with hidden type 'sgx_att_key_id_ext_t' returned by AesmClient::select_att_key_id or AesmKeyIds::select_algorithm_id or similar.
+    ///
+    /// @returns - information to use for get_quote_ex
+    ///
+    /// @note - intentionally skipping p_pub_key_id and buf_size as they are not useful at the moment.
+    ///         a) buf_size must be exactly sizeof(sgx_sha256_hash_t) otherwise fail.
+    ///         b) p_pub_key_id set to false causes response to be empty except 'pub_key_id_size'. But we must already know that and set it via buf_size - otherwise see point a.
+    ///
+    pub fn init_quote_ex(&self, att_key_id: Vec<u8>) -> Result<QuoteInfoEx> {
+
+        let mut req = Request_InitQuoteExRequest::new();
+
+        // FIXME: remove conditional compilation after resolving https://github.com/fortanix/rust-sgx/issues/31
+        #[cfg(not(target_env = "sgx"))]
+        req.set_timeout(REMOTE_AESM_TIMEOUT_US);
+
+        req.set_att_key_id(att_key_id.clone());
+        req.set_b_pub_key_id(true);
+        req.set_buf_size(::std::mem::size_of::<sgx_sha256_hash_t>() as u64);
+        
+        let mut res = self.transact(req)?;
+
+        Ok(QuoteInfoEx { att_key_id,
```

Comment:
I don't think we need to return `att_key_id`. The user passed it in already. (After you remove this: unnecessary clone above)

Reply:
This is a copy to make it easier for 'get_quote_ex' to work.

'get_quote_ex' also needs a key, either we put it as parameter to the function or we store it for convenience in the 'QuoteInfoEx'

Reply:
Why clone here if the caller can just as easily clone

Reply:
fixed as you mentioned - caller is now responsible for clone